### PR TITLE
verifying paymaster contract and tests

### PR DESCRIPTION
**Pull request by talhamalik883**

### CODE REVIEW

File: contracts/references/aa-4337/VerifyingPayMaster.sol [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+/**
+ * A sample paymaster that uses external service to decide whether to pay for the UserOp.
+ * The paymaster trusts an external signer to sign the transaction.
+ * The calling user must pass the UserOp to that external signer first, which performs
+ * whatever off-chain verification before signing the UserOp.
+ * Note that this signature is NOT a replacement for wallet signature:
+ * - the paymaster signs to agree to PAY for GAS.
+ * - the wallet signs to prove identity and wallet ownership.
+ */
+contract VerifyingPayMaster is BasePaymaster, Initializable {
```
Comment (livingrockrises) [inline contracts/references/aa-4337/VerifyingPayMaster.sol:20]: rename this to VerifyingPaymaster 

make sure everywhere it goes paymaster without M caps

File: contracts/smart-contract-wallet/VerifyPayMasterProxy.sol [hunk 2]
```diff
@@ -0,0 +1,40 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+/**
+ * @title Proxy // This is the user's wallet
+ * @notice Basic proxy that delegates all calls to a fixed implementing contract.
+ */
+contract VerifyPayMasterProxy {
```
Comment (livingrockrises) [inline contracts/smart-contract-wallet/VerifyPayMasterProxy.sol:8]: rename filename and within the contract 

with ...Paymaster..

File: contracts/smart-contract-wallet/VerifyPayMasterProxy.sol [hunk 3]
```diff
@@ -0,0 +1,40 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+/**
+ * @title Proxy // This is the user's wallet
```
Comment (livingrockrises) [inline contracts/smart-contract-wallet/VerifyPayMasterProxy.sol:5]: Review the comments. this is not the user's wallet

File: contracts/references/aa-4337/VerifyingPayMaster.sol [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        require(verifyingSigner == address(0), "Already initialized");
+        require(address(entryPoint) == address(0), "Already initialized");
+        verifyingSigner = _verifyingSigner;
+        entryPoint = _entryPoint;
+        owner = _owner;
+    }
+    /**
+    this function will let owner change signer
+    */
+    function setSigner( address _newSigner) external onlyOwner{
```
Comment (livingrockrises) [inline contracts/references/aa-4337/VerifyingPayMaster.sol:41]: can we make it _newVerifyingSigner

File: contracts/smart-contract-wallet/VerifyingPayMasterFactory.sol [hunk 5]
```diff
@@ -0,0 +1,28 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "./VerifyPayMasterProxy.sol";
+import "../references/aa-4337/VerifyingPayMaster.sol"; 
+
+contract VerifyingPayMasterFactory {
```
Comment (livingrockrises) [inline contracts/smart-contract-wallet/VerifyingPayMasterFactory.sol:7]: rename

File: contracts/smart-contract-wallet/VerifyingPayMasterFactory.sol [hunk 6]
```diff
@@ -0,0 +1,28 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "./VerifyPayMasterProxy.sol";
```
Comment (livingrockrises) [inline contracts/smart-contract-wallet/VerifyingPayMasterFactory.sol:4]: rename

File: contracts/smart-contract-wallet/VerifyingPayMasterFactory.sol [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+import "../references/aa-4337/VerifyingPayMaster.sol"; 
+
+contract VerifyingPayMasterFactory {
+    address internal payMasterImp; 
+
+    event VerifyingPayMasterCreated(address indexed proxy, address payMasterImp, IEntryPoint entryPoint, address indexed owner, address indexed verifyingSigner);
```
Comment (livingrockrises) [inline contracts/smart-contract-wallet/VerifyingPayMasterFactory.sol:10]: rename

File: contracts/smart-contract-wallet/VerifyingPayMasterFactory.sol (hunk 6 above)
Comment (livingrockrises) [inline contracts/smart-contract-wallet/VerifyingPayMasterFactory.sol:4]: shouldn't this be named VerifyingPaymasterProxy

File: scripts/smart-wallet/deployWallet.ts [hunk 8]
```diff
@@ -3,12 +3,13 @@
 //
 // When running the script with `npx hardhat run <script>` you'll find the Hardhat
```
Comment (livingrockrises) [inline scripts/smart-wallet/deployWallet.ts:4]: let's not call this script deployWallet.ts which was already existing for wallet depoyment

can create another one for paymaster deployment

File: scripts/smart-wallet/deployWallet.ts [hunk 9]
```diff
@@ -18,52 +19,82 @@ async function main() {
...
+    "VerifyingPayMasterFactory"
   );
-  const handler = await DefaultHandler.deploy();
-  await handler.deployed();
-  console.log("Default callback handler deployed at: ", handler.address);
+  const verifyingPayMasterFactory = await VerifyingPayMasterFactory.deploy(payMasterImp.address);
+  await verifyingPayMasterFactory.deployed();
+  console.log("verifyingPayMasterFactory Implementation deployed at: ", verifyingPayMasterFactory.address);
+  
+  await verifyingPayMasterFactory.deployVerifyingPayMaster('0xD68fdc0B89010a9039C2C38f4a3E5c4Ed98f7bC1', '0xD68fdc0B89010a9039C2C38f4a3E5c4Ed98f7bC1', '0x2167fA17BA3c80Adee05D98F0B55b666Be6829d6')
```
Comment (livingrockrises) [inline scripts/smart-wallet/deployWallet.ts:81]: let's not hard code addresses here. deploy required contracts and pass args

File: test/smart-wallet/testGroup1.ts [hunk 10]
```diff
@@ -121,7 +121,7 @@ describe("Base Wallet Functionality", function () {
       )
     )
       .to.emit(walletFactory, "WalletCreated")
-      .withArgs(expected, baseImpl.address, owner);
```
Comment (livingrockrises) [inline test/smart-wallet/testGroup1.ts:124]: search this in entire project and replace with additional event arguments

File: contracts/references/aa-4337/VerifyingPayMaster.sol (hunk 1 above)
Comment (talhamalik883) [reply contracts/references/aa-4337/VerifyingPayMaster.sol]: done

File: contracts/smart-contract-wallet/VerifyPayMasterProxy.sol (hunk 3 above)
Comment (talhamalik883) [reply contracts/smart-contract-wallet/VerifyPayMasterProxy.sol]: done

**Pull request merged**